OpenRA.Mods.D2k/Traits/SpiceBloom.cs

```diff
@@ -0,0 +1,148 @@
+
+		[Desc("The range of time (in ticks) that the spicebloom will take to grow.")]
+		public readonly int[] GrowthDelay = { 1000, 1500 };
+
+		public readonly string ResourceType = "Spice";
+
+		[Desc("The weapon to use for spice creation.")]
+		[WeaponReference]
+		public readonly string Weapon = "SpiceExplosion";
+
+		[Desc("The amount of spice to expel.")]
+		public readonly int[] Pieces = { 3, 10 };
+
+		[Desc("The maximum distance in cells that spice may be expelled.")]
+		public readonly int Range = 5;
+
+		public object Create(ActorInitializer init) { return new SpiceBloom(init, this); }
+	}
+
+	public class SpiceBloom : ITick, INotifyKilled
+	{
+		readonly Actor self;
+		readonly SpiceBloomInfo info;
+		readonly ResourceType resType;
+		readonly ResourceLayer resLayer;
+		readonly AnimationWithOffset anim;
+
+		readonly int respawnTicks;
+		readonly int growTicks;
+		int ticks;
+
+		public SpiceBloom(ActorInitializer init, SpiceBloomInfo info)
+		{
+			this.info = info;
+			self = init.Self;
+
+			resLayer = self.World.WorldActor.Trait<ResourceLayer>();
+			resType = self.World.WorldActor.TraitsImplementing<ResourceType>().First(t => t.Info.Name == info.ResourceType);
+
+			var render = self.Trait<RenderSprites>();
+			anim = new AnimationWithOffset(new Animation(init.Self.World, render.GetImage(self)), null, () => self.IsDead);
+			render.Add(anim);
+
+			respawnTicks = self.World.SharedRandom.Next(info.RespawnDelay[0], info.RespawnDelay[1]);
+			growTicks = self.World.SharedRandom.Next(info.GrowthDelay[0], info.GrowthDelay[1]);
+			anim.Animation.Play(info.GrowthSequences[0]);
+		}
+
+		public void Tick(Actor self)
+		{
+			ticks++;
+
+			if (ticks >= growTicks)
+				self.Kill(self);
+			else
+			{
+				var index = info.GrowthSequences.Length * ticks / growTicks;
+				anim.Animation.Play(info.GrowthSequences[index]);
+			}
+		}
+
+		public void Killed(Actor self, AttackInfo e)
+		{
+			var args = new ProjectileArgs
+			{
+				Weapon = self.World.Map.Rules.Weapons[info.Weapon.ToLowerInvariant()],
+				Facing = 0,
+
+				DamageModifiers = self.TraitsImplementing<IFirepowerModifier>()
+					.Select(a => a.GetFirepowerModifier()).ToArray(),
+
+				InaccuracyModifiers = self.TraitsImplementing<IInaccuracyModifier>()
+					.Select(a => a.GetInaccuracyModifier()).ToArray(),
+
+				Source = self.CenterPosition,
+				SourceActor = self,
+			};
+
+			var pieces = self.World.SharedRandom.Next(info.Pieces[0], info.Pieces[1]) * ticks / growTicks;
+			for (var i = 0; pieces > i; i++)
+			{
+				var cells = OpenRA.Traits.Util.RandomWalk(self.Location, self.World.SharedRandom);
+				var cell = cells.Take(info.Range).SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p)).Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+				if (cell == null)
+					cell = cells.Take(info.Range).Random(self.World.SharedRandom);
+
+				args.PassiveTarget = self.World.Map.CenterOfCell(cell.Value);
```

Comment:
Actually this is not the problem. This has to be set for each projectile. The randomization a few lines before just doesn't work, because `cell` is always null.